docs/building-in-chunks/circuits.md

```diff
@@ -0,0 +1,178 @@
+
+```mermaid
+graph TD
+    A[EpochRoot] --> B[CheckpointMerge]
+    A --> C[CheckpointMerge]
+
+    B --> D[CheckpointRoot]
+    B --> E[CheckpointRoot]
+
+    D --> F[BlockMerge]
+    D --> G[BlockMerge]
+
+    F --> H[BlockRoot]
+    F --> I[BlockRoot]
+
+    H --> J[TxMerge]
+    H --> K[TxMerge]
+
+    J --> L[BaseRollup]
+    J --> M[BaseRollup]
+
+    D --> P[ParityRoot]
+    P --> PB1[ParityBase]
+    P --> PB2[ParityBase]
+```
+
+All world state changes are executed in the `BlockRoot` circuits, and the `CheckpointRoot` circuits aggregate updates from multiple `BlockRoot`s. The `CheckpointRoot` then produces a `ContentCommitment` as we have today.
+
+#### L1-to-L2 messaging
+
+Assuming we keep the same strategy for consuming messages from the `Inbox` on every checkpoint to L1, then every chunk in a given checkpoint will have the same L1-to-L2 messages tree root. Today, this root is updated at the very end of the `BlockRoot` circuit. Under the new model, since the `CheckpointRoot` circuit should not be making changes to the world state trees, we have two options: either the first chunk of the checkpoint updates the tree before any tx effects, or the last chunk of the checkpoint updates it after having applied its tx effects.
+
+Depending on how we handle attestations for chunks, if we want the proposer to have the flexibility to decide on the fly which chunk will be the last one in the checkpoint, it's possible the proposer does not know which will be the last chunk. So it seems safer to update the L1-to-L2 message tree in the first chunk of a checkpoint.
+
+### L1 changes
+
+The main change for L1 contracts is that `propose` now advances the current block number by more than `1`. The rollup contract will need to store both the latest archive tree root, determined by the latest chunk included in the latest checkpoint, and the latest checkpoint hash. Since the checkpoint hash contains the latest archive tree root, it's possible we can just store the last checkpoint hash.
+
+## Fully decoupling L2 blocks from L1
+
+Should we ever want to decouple L2 block production from L1 checkpointing, there are a few changes needed to the design above. For clarity, by decoupling L2 blocks from L1 checkpointing, we mean having L2 produce blocks, and at some point an entity groups some of them into a checkpoint and pushes them to L1, but L2 can keep advancing independent of L1. In other words, L2 liveness should not depend on L1 availability (though its finality most definitely should).
+
+Since we do not know _when_ a chunk will make it to L1, we cannot know in which blob it will land, so we cannot know its `blobSponge` start and end state. One option here would be to just keep a rolling hash of `TxEffects` which gets permanently accumulated and is never reset, so when the network decides to aggregate a few chunks into a checkpoint, the blob verification circuits need to account for a non-zero initial sponge (I need to check feasibility of this).
```

Comment:
Very clever idea.
This should be possible. At the moment we assert that the "start" sponge of a block is empty. Instead, we can assert that the start sponge of a chunk is always that of the previous -- even for the 1st chunk in a checkpoint.